Bottom: d82222c2c78634352bb339ba72f39de5b2846159
Top:    082fa615b5ca842077f1b241b5c2d35387c545ec
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-25 17:37:04 +0100

draw.c: Split "Gathering" routines from real drawing routines (Lines)


---

diff --git a/src/draw.c b/src/draw.c
index 9051686..6a3b6f1 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -89,7 +89,6 @@ static const BoxType *clip_box = NULL;
 static void DrawEverything (BoxTypePtr);
 static void DrawPPV (int group, const BoxType *);
 static int DrawLayerGroup (int, const BoxType *);
-static void DrawLineLowLevel (LineTypePtr);
 static void DrawRegularText (LayerTypePtr, TextTypePtr);
 static void DrawPolygonLowLevel (PolygonTypePtr);
 static void DrawArcLowLevel (ArcTypePtr);
@@ -486,6 +485,42 @@ rat_callback (const BoxType * b, void *cl)
 }
 
 static void
+_draw_line (LineType *line)
+{
+  gui->set_line_cap (Output.fgGC, Trace_Cap);
+  if (TEST_FLAG (THINDRAWFLAG, PCB))
+    gui->set_line_width (Output.fgGC, 0);
+  else
+    gui->set_line_width (Output.fgGC, line->Thickness);
+
+  gui->draw_line (Output.fgGC,
+		  line->Point1.X, line->Point1.Y,
+		  line->Point2.X, line->Point2.Y);
+}
+
+static void
+draw_line (LayerType *layer, LineType *line)
+{
+  if (TEST_FLAG (SELECTEDFLAG | FOUNDFLAG, line))
+    {
+      if (TEST_FLAG (SELECTEDFLAG, line))
+        gui->set_color (Output.fgGC, layer->SelectedColor);
+      else
+        gui->set_color (Output.fgGC, PCB->ConnectedColor);
+    }
+  else
+    gui->set_color (Output.fgGC, layer->Color);
+  _draw_line (line);
+}
+
+static int
+line_callback (const BoxType * b, void *cl)
+{
+  draw_line ((LayerType *) cl, (LineType *) b);
+  return 1;
+}
+
+static void
 draw_element_package (ElementType *element)
 {
   /* set color and draw lines, arcs, text and pins */
@@ -501,7 +536,7 @@ draw_element_package (ElementType *element)
   /* draw lines, arcs, text and pins */
   ELEMENTLINE_LOOP (element);
   {
-    DrawLineLowLevel (line);
+    _draw_line (line);
   }
   END_LOOP;
   ARC_LOOP (element);
@@ -904,13 +939,6 @@ DrawRats (BoxTypePtr drawn_area)
 }
 
 static int
-line_callback (const BoxType * b, void *cl)
-{
-  DrawLine ((LayerTypePtr) cl, (LineTypePtr) b);
-  return 1;
-}
-
-static int
 arc_callback (const BoxType * b, void *cl)
 {
   DrawArc ((LayerTypePtr) cl, (ArcTypePtr) b);
@@ -1061,29 +1089,6 @@ GatherPadName (PadTypePtr Pad)
 }
 
 /* ---------------------------------------------------------------------------
- * lowlevel drawing routine for lines
- */
-static void
-DrawLineLowLevel (LineTypePtr Line)
-{
-  if (Gathering)
-    {
-      AddPart (Line);
-      return;
-    }
-
-  gui->set_line_cap (Output.fgGC, Trace_Cap);
-  if (TEST_FLAG (THINDRAWFLAG, PCB))
-    gui->set_line_width (Output.fgGC, 0);
-  else
-    gui->set_line_width (Output.fgGC, Line->Thickness);
-
-  gui->draw_line (Output.fgGC,
-		  Line->Point1.X, Line->Point1.Y,
-		  Line->Point2.X, Line->Point2.Y);
-}
-
-/* ---------------------------------------------------------------------------
  * lowlevel drawing routine for text objects
  */
 void
@@ -1137,7 +1142,7 @@ DrawTextLowLevel (TextTypePtr Text, int min_line_width)
 	      newline.Point1.Y += Text->Y;
 	      newline.Point2.X += Text->X;
 	      newline.Point2.Y += Text->Y;
-	      DrawLineLowLevel (&newline);
+	      _draw_line (&newline);
 	    }
 
 	  /* move on to next cursor position */
@@ -1293,19 +1298,9 @@ DrawPadName (PadTypePtr Pad)
 void
 DrawLine (LayerTypePtr Layer, LineTypePtr Line)
 {
-  if (!Gathering)
-    {
-      if (TEST_FLAG (SELECTEDFLAG | FOUNDFLAG, Line))
-	{
-	  if (TEST_FLAG (SELECTEDFLAG, Line))
-	    gui->set_color (Output.fgGC, Layer->SelectedColor);
-	  else
-	    gui->set_color (Output.fgGC, PCB->ConnectedColor);
-	}
-      else
-	gui->set_color (Output.fgGC, Layer->Color);
-    }
-  DrawLineLowLevel (Line);
+  assert (Gathering);
+
+  AddPart (Line);
 }
 
 /* ---------------------------------------------------------------------------
@@ -1354,7 +1349,7 @@ DrawRat (RatTypePtr Line)
 	}
     }
   else
-    DrawLineLowLevel ((LineTypePtr) Line);
+    _draw_line ((LineTypePtr) Line);
 }
 
 /* ---------------------------------------------------------------------------
@@ -1540,7 +1535,7 @@ DrawElementPackage (ElementTypePtr Element)
 
   ELEMENTLINE_LOOP (Element);
   {
-    DrawLineLowLevel (line);
+    DrawLine (NULL, line);
   }
   END_LOOP;
   ARC_LOOP (Element);
@@ -1602,7 +1597,7 @@ EraseRat (RatTypePtr Rat)
 		     w * 2, w * 2, 0, 360);
     }
   else
-    DrawLineLowLevel ((LineTypePtr) Rat);
+    _draw_line ((LineTypePtr) Rat);
 }
 
 
@@ -1671,7 +1666,9 @@ ErasePinName (PinTypePtr Pin)
 void
 EraseLine (LineTypePtr Line)
 {
-  DrawLineLowLevel (Line);
+  assert (Gathering);
+
+  AddPart (Line);
 }
 
 /* ---------------------------------------------------------------------------
@@ -1717,7 +1714,7 @@ EraseElement (ElementTypePtr Element)
 {
   ELEMENTLINE_LOOP (Element);
   {
-    DrawLineLowLevel (line);
+    EraseLine (line);
   }
   END_LOOP;
   ARC_LOOP (Element);
